File: 数据结构篇/单调栈/42. 接雨水.py
# 给定 n 个非负整数表示每个宽度为 1 的柱子的高度图，计算按此排列的柱子，下雨之后能接多少雨水。

# 输入：height = [0,1,0,2,1,0,1,3,2,1,2,1]
# 输出：6
# 解释：上面是由数组 [0,1,0,2,1,0,1,3,2,1,2,1] 表示的高度图，在这种情况下，可以接 6 个单位的雨水（蓝色部分表示雨水）。 
# 示例 2：

# 输入：height = [4,2,0,3,2,5]
# 输出：9
 

# 提示：

# n == height.length
# 1 <= n <= 2 * 104
# 0 <= height[i] <= 105

class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        
        
        stack=list()
        stack.append(0)
        
        for i in range(1,len(height)):
            if height[i]<stack[-1]:
                stack.append(i)
            elif height[i]==stack[-1]:
                stack.append(i)
            else:
                pass
                
        #双指针
        # sum=0
        # maxLeft=[0 for i in range(len(height))]
        # maxRight=[0 for i in range(len(height))]
        
        # size=len(height)
        # maxLeft[0]=height[0]
        # for i in range(1,len(height)):
        #     maxLeft[i]=max(maxLeft[i-1],height[i])
            
        # maxRight[size-1]=height[size-1]
        
        # for i in range(size-2,-1,-1):
        #     maxRight[i]=max(maxRight[i+1],height[i])
            
        # for i in range(size):
        #     count=min(maxLeft[i],maxRight[i])-height[i]
        #     if count>0:
        #         sum=sum+count
                
        # return sum
        
solution=Solution()
height=[4,2,0,3,2,5]
print(solution.trap(height))

File: 数据结构篇/双指针/42.接雨水.cpp


#include <vector>
#include <queue>
#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
    /* data */
};

class Solution
{
public:
    int trap(vector<int> &height)
    {
        if(height.size()<=2)
        {
            return 0;
        }
        vector<int> maxLeft(height.size(),0);
        vector<int> maxRight(height.size(),0);
        int size=maxRight.size();

        maxLeft[0]=height[0];
        for(int i=1;i<size;i++)
        {
            maxLeft[i]=max(maxLeft[i-1],height[i]);
        }  
        maxLeft[size-1]=height[size-1];
        for(int i=size-2;i>=0;i--)
        {
            maxRight[i]=max(maxRight[i+1],height[i]);
        }

        int sum=0;
        for(int i=0;i<size;i++)
        {
            int count=min(maxLeft[i],maxRight[i])-height[i];
            if(count>0)
            {
                sum+=count;
            }
        }
        return sum;
    }
private:
};

int main()
{
    vector<int> height = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}; // 示例输入
    Solution solution;
    cout << "Water trapped: " << solution.trap(height) << endl;
    return 0;
}